### ISSUE TITLE
Use with nodenv

### ISSUE THREAD

**Issue by tylercollier** [zkat/npx#82]
(I've migrated this issue from its original home on the [nodenv issues](https://github.com/nodenv/nodenv/issues/79), where I closed it, thinking this probably needs to be handled on the npx side)

Using the [nodenv node version manager](https://github.com/nodenv/nodenv), I want to use `$ npx create-react-app my-cool-new-app` where it installs a temporary create-react-app and calls it. So I installed node 8. I did `nodenv global 8.0.0`. I upgraded npm to make sure I had 5.2. I then ran:

```
npx create-react-app real-world-discussions-demo
```

Output:
```
nodenv: create-react-app: command not found

The `create-react-app' command exists in these Node versions:
  4.2.3
```

Oops! I had already installed `create-react-app` globally (with nodenv) using 4.2.3. But wait, I'm not trying to run a global command, right? That's the point of npx. After I deleted create-react-app out of 4.2.3, I ran the command again and it worked.

So my question is, why didn't this work as expected? It should have installed a temporary create-react-app in node 8 and called it, but instead it barfed, but that's exactly what npx is trying to prevent. The [original npx article](https://medium.com/@maybekatz/introducing-npx-an-npm-package-runner-55f7d4bd282b) mentions that npx should work well with nvm, nave, or n, but doesn't mention nodenv.

I already asked Kat about this [here](https://medium.com/@tylercollier_24031/awesome-297684337f75), and she responded saying:

> npx has no explicit support for any version managers. It does some installation and $PATH-manipulation work, and that’s it

I suspect npx would need to be aware of nodenv. My understanding is that nodenv uses shims in $PATH such that the system will use it first before global stuff, whereas the other node version managers modify things on a more global level. If I understand correctly, the other ones only allow you to have one node version active at a time. I could be wrong, but that sounds like it would account for the difference that this problem would not occur with the other version managers but would for nodenv.

So my feature request is: make npx even more awesome and support nodenv! I realize nodenv is less popular (_why??_), so I'm a diva here asking for this special case. No rush here for me; I've been using the workarounds mentioned in the article for some time :-) But, I just wanted to say thanks for making npx, it's a welcome tool!

**Comment by zkat** [comment]
I would love some help with this. I've no idea at all what nodenv is doing, and I'm sure whatever cleverness it's up to is clobbering npx's own PATH-munging. If someone takes time to do the research as to _why_ this is happening, I'd be more than happy to consider putting a patch together. No promises, but let's figure this out?

**Comment by tylercollier** [comment]
Arguably nodenv isn't clever at all. It just puts itself earlier in your PATH so that it's seen first before (e.g.) `/usr/local/bin`. Arguably the other environment managers need to be more clever. I think it's explained clearly at the top of their [README](https://github.com/nodenv/nodenv#how-it-works), and the maintainer Jason Karns mentioned what's happening specifically [here](https://github.com/nodenv/nodenv/issues/79#issuecomment-315075031), that npx is indeed finding something in the path, but only after it's invoked does nodenv say "well, it exists, but not in this version of node". So perhaps what's needed is on both sides: npx would need to ask "does this binary exist, and if so, does it exist for version X of node?" and nodenv would have to have a way to respond to that request. I realize this would be considered a special case, just for nodenv, and might be out of scope.

**Comment by zkat** [comment]
@tylercollier the only workaround for this I can think of is for nodenv itself to have just enough awareness of `$PATH` to recognize npx's temporary installs into `~/.npm/_npx/<pid>` prefixes. If a command exists, and nodenv then pretends it doesn't, that doesn't seem like npx's problem -- beyond the fact that those installs are temporary, they are for all intents and purposes treated like global installs by npm. nodenv's current model seems just plain incompatible with tools like npx.

How does nodenv interact with tools that add `<project>/node_modules/.bin` to `$PATH`? That's the same logic it should follow for npx, since that's another path that npx looks into.

Frankly, I'm inclined to close this, because I see nothing that npx can do here -- it would need to literally have chunks of its internals rewritten to _specifically target nodenv_, and even then _might not work if nodenv changes its shim mechanics_. I really really don't feel great with the outlook of that.

That said, if you have questions about how npx does its thing and you're interested in either patching nodenv or writing an npx wrapper tool with your custom logic, then feel free to ask questions about the internal mechanics of npx and I'll do my best to elucidate!

**Comment by zkat** [comment]
I guess I'll close this for now. I don't think there's anything to do npx-side that I think is reasonable for npx itself to do. If there's a more concrete fix that isn't completely intrusive, we can talk about more specific things.
